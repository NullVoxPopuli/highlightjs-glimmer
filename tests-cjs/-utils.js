// NOTE: this highlight.js was manually built and had an export manually added
//       This is because the glimmer syntax requires a to-be-released
//       version of highlight.js at the time of writing
import hljs from 'highlight.js';
import { registerLanguage } from 'highlightjs-glimmer';

registerLanguage(hljs);

export function parse(code, lang = 'glimmer') {
  return hljs.highlight(code, { language: lang }).value;
}

export function tag(klass, children = []) {
  let _children = Array.isArray(children) ? children : [children];

  return `<span class="hljs-${klass}">${_children.join('')}</span>`;
}
