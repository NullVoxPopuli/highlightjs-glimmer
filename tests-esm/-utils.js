import { expect } from 'vitest';
import hljs from 'highlight.js';
import { setup } from 'highlightjs-glimmer';

setup(hljs);

import prettier from 'prettier';

// Used to resolve indentation issues.
// We don't care about that in these tests -- we only care about tag identification.
export function format(code) {
  return prettier.format(code, { parser: 'html', htmlWhitespaceSensitivity: 'ignore' });
}

export function formattedEquals(actual, expected) {
  expect(format(actual)).toEqual(format(expected));
}

export function parse(code, lang = 'glimmer') {
  return hljs.highlight(code, { language: lang }).value;
}

export function glimmer(...children) {
  return `<span class="language-glimmer">${children.join('')}</span>`;
}

export const templateTags = {
  get begin() {
    return tag('tag', ['&lt;', tag('title', 'template'), '&gt;']);
  },
  get end() {
    return tag('tag', ['&lt;/', tag('title', 'template'), '&gt;']);
  },
};

export function template(...content) {
  return list(templateTags.begin, '\n', ...indentLines(content), '\n', templateTags.end);
}

export function selfClosing(name, attributes = []) {
  return tag('tag', ['&lt;', tag('title', name), ...attributes, ' /&gt;']);
}

export function indentLines(lines, amount = 2) {
  let indent = Array(amount + 1).join(' ');

  return lines.map((line) => {
    return `${indent}${line}`;
  });
}

export function element(name, content = []) {
  return list(
    tag('tag', ['&lt;', tag('title', name), '&gt;']),
    '\n',
    ...indentLines(content),
    '\n',
    tag('tag', ['&lt;/', tag('title', name), '&gt;'])
  );
}

export function mustache(...content) {
  return tag('punctuation mustache', ['{{', ...content, '}}']);
}

export function arg(name) {
  return list(tag('punctuation', '@'), tag('params', name));
}

export const keyword = {
  export: tag('keyword', 'export'),
  const: tag('keyword', 'const'),
  default: tag('keyword', 'default'),
};

export const operator = {
  equals: tag('operator', '='),
};

export function string(content) {
  return tag('string', ['&quot;', content, '&quot;']);
}

export const tags = {
  template,
  mustache,
  element,
  selfClosing,
  arg,
  keyword,
  operator,
  string,
};

export function tag(klass, children = []) {
  let _children = Array.isArray(children) ? children : [children];

  return `<span class="hljs-${klass}">${_children.join('')}</span>`;
}

export function list(...children) {
  return children.join('');
}
