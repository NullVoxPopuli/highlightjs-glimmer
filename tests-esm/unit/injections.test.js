import { describe, test, expect } from 'vitest';
import { stripIndent } from 'common-tags';
import { parse, tag, list, glimmer, template, tags, formattedEquals, format } from '../-utils';

describe('Injections | JS', () => {
  describe('hbs template literal', () => {
    test('basic', () => {
      let result = parse(
        stripIndent`
          const hbs\`{{foo}}\`
        `,
        'js'
      );

      expect(result).toEqual(
        '<span class="hljs-keyword">const</span> hbs`<span class="language-glimmer"><span class="hljs-punctuation mustache">{{<span class="hljs-title">foo</span>}}</span>`</span>'
      );
    });
  });

  describe('gjs / template tag', () => {
    test('basic', () => {
      let result = parse(
        stripIndent`
          export const Name = <template>
            {{@name}}
          </template>;
        `,
        'glimmer-javascript'
      );

      expect(result).toEqual(
        list(
          tags.keyword.export,
          ' ',
          tags.keyword.const,
          ' ',
          tag('title class_', 'Name'),
          ' = ',
          glimmer(template(tags.mustache(tags.arg('name')))),
          ';'
        )
      );
    });

    test('implied default export', () => {
      let result = parse(
        stripIndent`
          <template>
            {{@name}}
          </template>
        `,
        'js'
      );

      expect(result).toEqual(glimmer(template(tags.mustache(tags.arg('name')))));
    });

    test('explicit default export', () => {
      let result = parse(
        stripIndent`
          export default <template>
            {{@name}}
          </template>
        `,
        'javascript'
      );

      expect(result).toEqual(
        list(
          tags.keyword.export,
          ' ',
          tags.keyword.default,
          ' ',
          glimmer(template(tags.mustache(tags.arg('name'))))
        )
      );
    });

    test('with imports', () => {
      let result = parse(
        stripIndent`
          import Greeting from './greeting.js';
          import WeatherSummary from './weather-summary.js';

          <template>
            <div>
              <Greeting @name="Chris" />
              <WeatherSummary />
            </div>
          </template>
        `,
        'js'
      );

      formattedEquals(
        result,
        list(
          // These are JS and not tagged by us
          `<span class="hljs-keyword">import</span> <span class="hljs-title class_">Greeting</span> <span class="hljs-keyword">from</span> <span class="hljs-string">&#x27;./greeting.js&#x27;</span>;`,
          '\n',
          `<span class="hljs-keyword">import</span> <span class="hljs-title class_">WeatherSummary</span> <span class="hljs-keyword">from</span> <span class="hljs-string">&#x27;./weather-summary.js&#x27;</span>;`,
          '\n',
          '\n',
          glimmer(
            template(
              tags.element('div', [
                tags.selfClosing('Greeting', [
                  ' ',
                  tags.arg('name'),
                  tags.operator.equals,
                  tags.string('Chris'),
                ]),
                '\n',
                tags.selfClosing('WeatherSummary'),
              ])
            )
          )
        )
      );
    });

    test('a function exists above the template', () => {
      let result = parse(
        stripIndent`
          import Greeting from './greeting.js';
          import WeatherSummary from './weather-summary.js';

          function isBirthday(dateOfBirth) {
            const now = new Date();
            return (
              dateOfBirth.getDate() === now.getDate() &&
              dateOfBirth.getMonth() === now.getMonth()
            );
          }

          <template>
            <div>
              <Greeting @name="Chris" />
              {{#if (isBirthday @user.dateOfBirth)}}
                <Celebration type='birthday' />
              {{/if}}
              <WeatherSummary />
            </div>
          </template>
        `,
        'js'
      );

      formattedEquals(
        result,
        list(
          // These are JS and not tagged by us
          `<span class="hljs-keyword">import</span> <span class="hljs-title class_">Greeting</span> <span class="hljs-keyword">from</span> <span class="hljs-string">&#x27;./greeting.js&#x27;</span>;
 <span class="hljs-keyword">import</span> <span class="hljs-title class_">WeatherSummary</span> <span class="hljs-keyword">from</span> <span class="hljs-string">&#x27;./weather-summary.js&#x27;</span>;

 <span class="hljs-keyword">function</span> <span class="hljs-title function_">isBirthday</span>(<span class="hljs-params">dateOfBirth</span>) {
   <span class="hljs-keyword">const</span> now = <span class="hljs-keyword">new</span> <span class="hljs-title class_">Date</span>();
   <span class="hljs-keyword">return</span> (
     dateOfBirth.<span class="hljs-title function_">getDate</span>() === now.<span class="hljs-title function_">getDate</span>() &amp;&amp;
     dateOfBirth.<span class="hljs-title function_">getMonth</span>() === now.<span class="hljs-title function_">getMonth</span>()
   );
 }`,
          '\n',
          '\n',
          glimmer(
            template(
              tags.element('div', [
                tags.selfClosing('Greeting', [
                  ' ',
                  tags.arg('name'),
                  tags.operator.equals,
                  tags.string('Chris'),
                ]),
                '\n',
                tags.mustache(
                  '#',
                  tag('title', tag('built_in', 'if')),
                  tag('punctuation', '('),
                  tag('title', 'isBirthday'),
                  ' ',
                  tag('punctuation', '@'),
                  tag('params', 'user'),
                  tag('punctuation', '.'),
                  tag('title', 'dateOfBirth'),
                  tag('punctuation', ')')
                ),
                tags.selfClosing('Celebration', [
                  tag('attribute', 'type'),
                  tags.operator.equals,
                  tag('string', ['&#x27;', 'birthday', '&#x27;']),
                ]),
                tags.mustache('/', tag('title', tag('built_in', 'if'))),
                tags.selfClosing('WeatherSummary'),
              ])
            )
          )
        )
      );
    });

    test('is embedded in a class', () => {
      let result = parse(
        stripIndent`
          import Component from '@glimmer/component';
          import { gt, lt } from '@glimmer/helper';

          export default class WeatherSummary extends Component {
            @tracked currentTemp;

            interval;

            getWeather = () => {
              this.currentTemp = // something
            }

            constructor(owner, args) {
              super(owner, args);
              this.interval = setInterval(this.getWeather, 10000);
            }

            willDestroy() {
              super.willDestroy();
              clearInterval(this.interval);
            }

            <template>
              <p>
                The current temperature is {{this.currentTemp}}!
                {{#if (lt 50 this.currentTemp)}}
                  Brr! ðŸ¥¶
                {{else if (gt 80 this.currentTemp)}}
                  Yikes! ðŸ¥µ
                {{/if}}
              </p>
            </template>
          }
        `,
        'js'
      );

      expect(format(result)).toMatchSnapshot();
    });

    test('multiple components', () => {
      let result = parse(
        stripIndent`
          import WeatherSummary from './weather-summary.js';

          const Greeting = <template>
            <p>Hello, {{@name}}!</p>
          </template>;

          function isBirthday(dateOfBirth) {
            const now = new Date();
            return (
              dateOfBirth.getDate() === now.getDate() &&
              dateOfBirth.getMonth() === now.getMonth()
            );
          }

          <template>
            <div>
              <Greeting @name="Chris" />
              {{#if (isBirthday @user.dateOfBirth)}}
                <Celebration type='birthday' />
              {{/if}}
              <WeatherSummary />
            </div>
          </template>
        `,
        'js'
      );

      expect(format(result)).toMatchSnapshot();
    });
  });
});
