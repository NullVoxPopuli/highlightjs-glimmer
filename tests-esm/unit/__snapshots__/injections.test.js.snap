// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Injections | JS > gjs / template tag > a setTimeout exists above the template 1`] = `
"const demo <span class=\\"hljs-operator\\">=</span> new Demo();

setTimeout(() <span class=\\"hljs-operator\\">=</span>&gt; { 
  demo.theProperty <span class=\\"hljs-operator\\">=</span> 2; 
}, 500);

<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-title\\">template</span>&gt;</span>
  <span class=\\"hljs-punctuation mustache\\">{{<span class=\\"hljs-title\\">demo</span><span class=\\"hljs-punctuation\\">.</span><span class=\\"hljs-title\\">theProperty</span>}}</span>
<span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-title\\">tepmlate</span>&gt;</span>"
`;

exports[`Injections | JS > gjs / template tag > is embedded in a class 1`] = `
"<span class=\\"hljs-keyword\\">import</span>
<span class=\\"hljs-title class_\\">Component</span>
<span class=\\"hljs-keyword\\">from</span>
<span class=\\"hljs-string\\">&#x27;@glimmer/component&#x27;</span>
;
<span class=\\"hljs-keyword\\">import</span>
{ gt, lt }
<span class=\\"hljs-keyword\\">from</span>
<span class=\\"hljs-string\\">&#x27;@glimmer/helper&#x27;</span>
;

<span class=\\"hljs-keyword\\">export</span>
<span class=\\"hljs-keyword\\">default</span>
<span class=\\"hljs-keyword\\">class</span>
<span class=\\"hljs-title class_\\">WeatherSummary</span>
<span class=\\"hljs-keyword\\">extends</span>
<span class=\\"hljs-title class_ inherited__\\">Component</span>
{ @tracked currentTemp; interval; getWeather =
<span class=\\"hljs-function\\">() =&gt;</span>
{
<span class=\\"hljs-variable language_\\">this</span>
.
<span class=\\"hljs-property\\">currentTemp</span>
=
<span class=\\"hljs-comment\\">// something</span>
}

<span class=\\"hljs-title function_\\">constructor</span>
(
<span class=\\"hljs-params\\">owner, args</span>
) {
<span class=\\"hljs-variable language_\\">super</span>
(owner, args);
<span class=\\"hljs-variable language_\\">this</span>
.
<span class=\\"hljs-property\\">interval</span>
=
<span class=\\"hljs-built_in\\">setInterval</span>
(
<span class=\\"hljs-variable language_\\">this</span>
.
<span class=\\"hljs-property\\">getWeather</span>
,
<span class=\\"hljs-number\\">10000</span>
); }

<span class=\\"hljs-title function_\\">willDestroy</span>
(
<span class=\\"hljs-params\\"></span>
) {
<span class=\\"hljs-variable language_\\">super</span>
.
<span class=\\"hljs-title function_\\">willDestroy</span>
();
<span class=\\"hljs-built_in\\">clearInterval</span>
(
<span class=\\"hljs-variable language_\\">this</span>
.
<span class=\\"hljs-property\\">interval</span>
); }

<span class=\\"language-glimmer\\">
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">p</span>
    &gt;
  </span>
  The current temperature is
  <span class=\\"hljs-punctuation mustache\\">
    {{
    <span class=\\"hljs-class\\">this</span>
    <span class=\\"hljs-punctuation\\">.</span>
    <span class=\\"hljs-property\\">currentTemp</span>
    }}
  </span>
  !
  <span class=\\"hljs-punctuation mustache\\">
    {{#
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">if</span></span>
    <span class=\\"hljs-punctuation\\">(</span>
    <span class=\\"hljs-title\\">lt</span>
    <span class=\\"hljs-number\\">50</span>
    <span class=\\"hljs-class\\">this</span>
    <span class=\\"hljs-punctuation\\">.</span>
    <span class=\\"hljs-property\\">currentTemp)</span>
    }}
  </span>
  Brr! ðŸ¥¶
  <span class=\\"hljs-punctuation mustache\\">
    {{
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">else</span></span>
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">if</span></span>
    <span class=\\"hljs-punctuation\\">(</span>
    <span class=\\"hljs-title\\"><span class=\\"hljs-function\\">gt</span></span>
    <span class=\\"hljs-number\\">80</span>
    <span class=\\"hljs-class\\">this</span>
    <span class=\\"hljs-punctuation\\">.</span>
    <span class=\\"hljs-property\\">currentTemp)</span>
    }}
  </span>
  Yikes! ðŸ¥µ
  <span class=\\"hljs-punctuation mustache\\">
    {{/
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">if</span></span>
    }}
  </span>
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">p</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
</span>
}
"
`;

exports[`Injections | JS > gjs / template tag > multiple components 1`] = `
"<span class=\\"hljs-keyword\\">import</span>
<span class=\\"hljs-title class_\\">WeatherSummary</span>
<span class=\\"hljs-keyword\\">from</span>
<span class=\\"hljs-string\\">&#x27;./weather-summary.js&#x27;</span>
;

<span class=\\"hljs-keyword\\">const</span>
<span class=\\"hljs-title class_\\">Greeting</span>
=
<span class=\\"language-glimmer\\">
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">p</span>
    &gt;
  </span>
  Hello,
  <span class=\\"hljs-punctuation mustache\\">
    {{
    <span class=\\"hljs-punctuation\\">@</span>
    <span class=\\"hljs-params\\">name</span>
    }}
  </span>
  !
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">p</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
</span>
;

<span class=\\"hljs-keyword\\">function</span>
<span class=\\"hljs-title function_\\">isBirthday</span>
(
<span class=\\"hljs-params\\">dateOfBirth</span>
) {
<span class=\\"hljs-keyword\\">const</span>
now =
<span class=\\"hljs-keyword\\">new</span>
<span class=\\"hljs-title class_\\">Date</span>
();
<span class=\\"hljs-keyword\\">return</span>
( dateOfBirth.
<span class=\\"hljs-title function_\\">getDate</span>
() === now.
<span class=\\"hljs-title function_\\">getDate</span>
() &amp;&amp; dateOfBirth.
<span class=\\"hljs-title function_\\">getMonth</span>
() === now.
<span class=\\"hljs-title function_\\">getMonth</span>
() ); }

<span class=\\"language-glimmer\\">
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">div</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">Greeting</span>
    <span class=\\"hljs-punctuation\\">@</span>
    <span class=\\"hljs-params\\">name</span>
    <span class=\\"hljs-operator\\">=</span>
    <span class=\\"hljs-string\\">&quot;Chris&quot;</span>
    /&gt;
  </span>
  <span class=\\"hljs-punctuation mustache\\">
    {{#
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">if</span></span>
    <span class=\\"hljs-punctuation\\">(</span>
    <span class=\\"hljs-title\\">isBirthday</span>
    <span class=\\"hljs-punctuation\\">@</span>
    <span class=\\"hljs-params\\">user</span>
    <span class=\\"hljs-punctuation\\">.</span>
    <span class=\\"hljs-title\\">dateOfBirth</span>
    <span class=\\"hljs-punctuation\\">)</span>
    }}
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">Celebration</span>
    <span class=\\"hljs-attribute\\">type</span>
    <span class=\\"hljs-operator\\">=</span>
    <span class=\\"hljs-string\\">&#x27;birthday&#x27;</span>
    /&gt;
  </span>
  <span class=\\"hljs-punctuation mustache\\">
    {{/
    <span class=\\"hljs-title\\"><span class=\\"hljs-built_in\\">if</span></span>
    }}
  </span>
  <span class=\\"hljs-tag\\">
    &lt;
    <span class=\\"hljs-title\\">WeatherSummary</span>
    /&gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">div</span>
    &gt;
  </span>
  <span class=\\"hljs-tag\\">
    &lt;/
    <span class=\\"hljs-title\\">template</span>
    &gt;
  </span>
</span>
"
`;
